e2e/claim_test.go

```diff
@@ -0,0 +1,256 @@
+	require.NotEmpty(t, claimAddress)
+
+	// claim airdrop
+	claimMsgTemplate := `
+		{
+		"claim_mint_nft" : {
+				"minter_address":"%s"
+			}
+		}
+	`
+	claimMsgRaw := []byte(fmt.Sprintf(claimMsgTemplate, minterAddress))
+	_, err = msgServer.ExecuteContract(sdk.WrapSDKContext(ctx), &wasmtypes.MsgExecuteContract{
+		Contract: claimAddress,
+		Sender:   accs[1].Address.String(),
+		Msg:      claimMsgRaw,
+	})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "address is not allowed to claim")
+
+	// allow contract
+
+	app.ClaimKeeper.SetParams(ctx, claimtypes.Params{
+		AirdropEnabled:     true,
+		AirdropStartTime:   startDateTime,
+		DurationUntilDecay: claimtypes.DefaultDurationUntilDecay,
+		DurationOfDecay:    claimtypes.DefaultDurationOfDecay,
+		ClaimDenom:         claimtypes.DefaultClaimDenom,
+		AllowedClaimers: []claimtypes.ClaimAuthorization{
+			{
+				ContractAddress: claimAddress,
+				Action:          claimtypes.ActionBidNFT,
+			},
+		},
+	})
+
+	claimMsgRaw = []byte(fmt.Sprintf(claimMsgTemplate, minterAddress))
+	_, err = msgServer.ExecuteContract(sdk.WrapSDKContext(ctx), &wasmtypes.MsgExecuteContract{
+		Contract: claimAddress,
+		Sender:   accs[1].Address.String(),
+		Msg:      claimMsgRaw,
+	})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "address is not allowed to claim")
+	claim, err := app.ClaimKeeper.GetClaimRecord(ctx, addr1)
+	require.NoError(t, err)
+	require.False(t, claim.ActionCompleted[claimtypes.ActionMintNFT])
+
+	app.ClaimKeeper.SetParams(ctx, claimtypes.Params{
+		AirdropEnabled:     true,
+		AirdropStartTime:   startDateTime,
+		DurationUntilDecay: claimtypes.DefaultDurationUntilDecay,
+		DurationOfDecay:    claimtypes.DefaultDurationOfDecay,
+		ClaimDenom:         claimtypes.DefaultClaimDenom,
+		AllowedClaimers: []claimtypes.ClaimAuthorization{
+			{
+				ContractAddress: claimAddress,
+				Action:          claimtypes.ActionMintNFT,
+			},
+		},
+	})
+
+	balance := app.BankKeeper.GetBalance(ctx, accs[1].Address, "ustars")
+	require.Equal(t,
+		"1900000000",
+		balance.Amount.String(),
+	)
+
+	claimMsgRaw = []byte(fmt.Sprintf(claimMsgTemplate, minterAddress))
+	_, err = msgServer.ExecuteContract(sdk.WrapSDKContext(ctx), &wasmtypes.MsgExecuteContract{
+		Contract: claimAddress,
+		Sender:   accs[1].Address.String(),
+		Msg:      claimMsgRaw,
+	})
+	require.NoError(t, err)
+	balance = app.BankKeeper.GetBalance(ctx, accs[1].Address, "ustars")
+	perAction := claimRecords[0].InitialClaimableAmount.AmountOf(claimtypes.DefaultClaimDenom).Quo(sdk.NewInt(int64(len(claimRecords[0].ActionCompleted))))
+
+	require.Greater(t, perAction.Int64(), int64(0))
+	expectedBalance := perAction.Add(sdk.NewInt(1_900_000_000)) // user already had 19000
+	require.Equal(t,
+		expectedBalance.String(),
+		balance.Amount.String(),
+	)
+
+	claim, err = app.ClaimKeeper.GetClaimRecord(ctx, addr1)
+	require.NoError(t, err)
+	require.True(t, claim.ActionCompleted[claimtypes.ActionMintNFT])
```

Comment:
How/where is this data stored? Does this prevents an address from double claiming?

Reply:
yup everything is taken care from the native claim module, see final test is attempting to claim again but balance doesn't move.